### PR TITLE
[PW-2489] Fix afterpay component

### PR DESCRIPTION
**Pull request by AlexandrosMor**
**Description**
The latest checkout component doesn't support the dynamic field rendering so now we show all the personalDetails and billingAddress fields. In the billingaddress the country code drop down is empty. Added the country code and prefill the values in afterpay to create the request

**Tested scenarios**

**Fixed issue**:

### CODE REVIEW

File: Observer/AdyenAbstractDataAssignObserver.php [hunk 1]
```diff
@@ -59,11 +69,6 @@ protected function getValidatedAdditionalData(DataObject $data)
             $stateData = json_decode($additionalData[self::STATE_DATA], true);
         }
 
-        // Get validated state data array
-        if (!empty($stateData)) {
-            $stateData = $this->getArrayOnlyWithApprovedKeys($stateData, $this->approvedStateDataKeys);
-        }
-
```
Comment (cyattilakiss) [inline Observer/AdyenAbstractDataAssignObserver.php:66]: why is this removed?

File: Helper/Requests.php [hunk 2]
```diff
@@ -127,18 +125,39 @@ public function buildCustomerData(
...
                     $request['shopperName']['lastName'] = $lastName;
                 }
+
+                if ($dateOfBirth = $billingAddress->getDateOfBirth()) {
+                    $request['paymentMethod']['personalDetails']['dateOfBirth'] = $dateOfBirth;
+                }
+            }
+
+            //if payment method is afterpay_default override the parameters with the provided ones
+            if ($paymentMethod == 'afterpay_default') {
```
Comment (cyattilakiss) [inline Helper/Requests.php:155]: should be done for not just afterpay_default, if these values are sent from the frontend we can use them instead of the billing address values

File: Observer/AdyenHppDataAssignObserver.php [hunk 3]
```diff
@@ -57,9 +75,10 @@ public function execute(Observer $observer)
     {
         // Get request fields
         $data = $this->readDataArgument($observer);
-
         $additionalData = $this->getValidatedAdditionalData($data);
-
+        $this->adyenHelper->adyenLogger->addAdyenDebug("AdyenHppDataAssignObserver--AdditionalData" . json_encode($additionalData));
```
Comment (cyattilakiss) [inline Observer/AdyenHppDataAssignObserver.php]:
```suggestion
```

File: Observer/AdyenHppDataAssignObserver.php [hunk 4]
```diff
@@ -57,9 +75,10 @@ public function execute(Observer $observer)
...
+        $additionalData[self::STATE_DATA] = $this->checkoutStateDataValidator->getValidatedAdditionalData($additionalData[self::STATE_DATA]);
+        $this->adyenHelper->adyenLogger->addAdyenDebug("AdyenHppDataAssignObserver--AdditionalData--After" . json_encode($additionalData));
```
Comment (cyattilakiss) [inline Observer/AdyenHppDataAssignObserver.php]:
```suggestion
```

File: Observer/AdyenHppDataAssignObserver.php [hunk 5]
```diff
@@ -57,9 +75,10 @@ public function execute(Observer $observer)
     {
         // Get request fields
         $data = $this->readDataArgument($observer);
-
         $additionalData = $this->getValidatedAdditionalData($data);
-
+        $this->adyenHelper->adyenLogger->addAdyenDebug("AdyenHppDataAssignObserver--AdditionalData" . json_encode($additionalData));
+        $additionalData[self::STATE_DATA] = $this->checkoutStateDataValidator->getValidatedAdditionalData($additionalData[self::STATE_DATA]);
```
Comment (cyattilakiss) [inline Observer/AdyenHppDataAssignObserver.php]: Have you tested it with an HPP payment method which doesn't send the $additionalData[self::STATE_DATA] from the frontend? I believe it will throw a php error if it's missing.

File: view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js [hunk 6]
```diff
@@ -212,26 +212,57 @@ define(
...
+                            city = quote.shippingAddress().city;
+                            country = quote.shippingAddress().countryId;
+                            postalCode = quote.shippingAddress().postcode;
+                            street = quote.shippingAddress().street.join(" ");
+                            firstName = quote.shippingAddress().firstname;
+                            lastName = quote.shippingAddress().lastname;
+                            telephone = quote.shippingAddress().telephone;
+                            email = customerData.email;
+                            shopperGender = customerData.gender;
+                            shopperDateOfBirth = customerData.dob;
```
Comment (cyattilakiss) [inline view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js:232]: have you tested it with guest user?

File: Helper/Requests.php (hunk 2 above)
Comment (AlexandrosMor) [reply Helper/Requests.php:155]: I am gonna integrate the if statements with the existing ones

File: Gateway/Request/CheckoutDataBuilder.php [hunk 7]
```diff
@@ -86,12 +86,12 @@ public function build(array $buildSubject)
         $componentStateData = $payment->getAdditionalInformation(AdyenCcDataAssignObserver::STATE_DATA);
         $requestBody = array_merge($requestBody, $componentStateData);
 
-
         if (empty($requestBody['paymentMethod']['type']) && !empty(
             $payment->getAdditionalInformation(
                 AdyenHppDataAssignObserver::BRAND_CODE
             )
             )) {
+
```
Comment (Aleffio) [inline Gateway/Request/CheckoutDataBuilder.php:94]:
```suggestion
```
I think travis is complaining for this blank line

File: Observer/AdyenHppDataAssignObserver.php [hunk 8]
```diff
@@ -23,16 +23,27 @@
...
-     * @var array
+     * @var \Adyen\Service\Validator\CheckoutStateDataValidator
      */
+    public $checkoutStateDataValidator;
+
+    public $adyenHelper;
+
+    public function __construct(
+        \Adyen\Service\Validator\CheckoutStateDataValidator $checkoutStateDataValidator,
+        \Adyen\Payment\Helper\Data $adyenHelper
```
Comment (Aleffio) [inline Observer/AdyenHppDataAssignObserver.php:40]:
```suggestion
        \Adyen\Service\Validator\CheckoutStateDataValidator $checkoutStateDataValidator
```
please remove $adyenHelper if not used

File: view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js (hunk 6 above)
Comment (AlexandrosMor) [reply view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js:232]: The parameters are passing correct from the frontend to backend with the guest user

File: Helper/Requests.php [hunk 9]
```diff
@@ -85,69 +101,53 @@ public function buildCustomerData(
...
         if ($customerId > 0) {
             $request['shopperReference'] = $customerId;
+
+            $customer = $this->customerRepository->getById($customerId);
+
+            if (empty($request[self::DATE_OF_BIRTH]) && $dateOfBirth = $customer->getDob()) {
+                $request[self::DATE_OF_BIRTH] = $dateOfBirth;
+            }
+
+            //TODO check if Gender can be retrieved
```
Comment (Aleffio) [inline Helper/Requests.php:111]: is the todo still applicable?

Comment (cyattilakiss) [reply Helper/Requests.php:111]: not for this ticket but for future improvements
